File: go/engine/runtime.go
package engine

import (
	"fmt"
)

// WorkflowFunc defines the signature for workflow code.
type WorkflowFunc func(ctx *WorkflowContext, input map[string]interface{}) (interface{}, error)

// ActivityOptions configures activity execution.
type ActivityOptions struct {
	RetryPolicy                   *RetryPolicy
	ScheduleToCloseTimeoutSeconds int
	StartToCloseTimeoutSeconds    int
	HeartbeatTimeoutSeconds       int
}

// WorkflowContext provides deterministic APIs for workflow code.
type WorkflowContext struct {
	runner *WorkflowRunner
}

// ExecuteActivity schedules or replays an activity deterministically.
func (ctx *WorkflowContext) ExecuteActivity(name string, input map[string]interface{}, opts *ActivityOptions) (interface{}, error) {
	return ctx.runner.handleActivity(name, input, opts)
}

// Sleep schedules or replays a timer deterministically.
func (ctx *WorkflowContext) Sleep(durationSeconds int) error {
	_, err := ctx.runner.handleTimer(durationSeconds)
	return err
}

// WaitSignal blocks until a signal is available.
func (ctx *WorkflowContext) WaitSignal(name string) (map[string]interface{}, error) {
	return ctx.runner.handleSignal(name)
}

// GetVersion records or replays a version marker for deterministic upgrades.
func (ctx *WorkflowContext) GetVersion(changeID string, minSupported, maxSupported int) (int, error) {
	return ctx.runner.handleVersion(changeID, minSupported, maxSupported)
}

// Command represents a pending decision from workflow execution.
type Command struct {
	CommandType string
	Attributes  map[string]interface{}
}

// WorkflowRunResult is the outcome of a workflow task.
type WorkflowRunResult struct {
	Status  string
	Result  interface{}
	Error   string
	Command *Command
}

// ActivityError is returned when an activity fails.
type ActivityError struct {
	Message string
	Details map[string]interface{}
}

func (e ActivityError) Error() string {
	return e.Message
}

type workflowYield struct {
	status  string
	command *Command
}

// HistoryView is a preprocessed summary of history for deterministic replay.
type HistoryView struct {
	ActivityScheduled []HistoryEvent
	ActivityResults   map[string]HistoryEvent
	TimerScheduled    []HistoryEvent
	TimerResults      map[string]HistoryEvent
	Signals           map[string][]map[string]interface{}
	VersionMarkers    map[string]int
}

func NewHistoryView(history []HistoryEvent) HistoryView {
	view := HistoryView{
		ActivityResults: make(map[string]HistoryEvent),
		TimerResults:    make(map[string]HistoryEvent),
		Signals:         make(map[string][]map[string]interface{}),
		VersionMarkers:  make(map[string]int),
	}
	for _, event := range history {
		switch event.EventType {
		case EventActivityTaskScheduled:
			view.ActivityScheduled = append(view.ActivityScheduled, event)
		case EventActivityTaskCompleted, EventActivityTaskFailed, EventActivityTaskTimedOut:
			if id, ok := event.Attributes["activity_id"].(string); ok {
				view.ActivityResults[id] = event
			}
		case EventTimerScheduled:
			view.TimerScheduled = append(view.TimerScheduled, event)
		case EventTimerFired:
			if id, ok := event.Attributes["timer_id"].(string); ok {
				view.TimerResults[id] = event
			}
		case EventSignalReceived:
			name, _ := event.Attributes["signal_name"].(string)
			payload, _ := event.Attributes["payload"].(map[string]interface{})
			if name != "" {
				view.Signals[name] = append(view.Signals[name], payload)
			}
		case EventVersionMarkerRecorded:
			changeID, _ := event.Attributes["change_id"].(string)
			versionFloat, _ := event.Attributes["version"].(float64)
			if changeID != "" {
				view.VersionMarkers[changeID] = int(versionFloat)
			}
		}
	}
	return view
}

// WorkflowRunner deterministically executes workflow code.
type WorkflowRunner struct {
	workflowFunc       WorkflowFunc
	inputPayload       map[string]interface{}
	historyView        HistoryView
	activityIndex      int
	timerIndex         int
	signalConsumed     map[string]int
	defaultRetryPolicy RetryPolicy
}

func NewWorkflowRunner(workflowFunc WorkflowFunc, inputPayload map[string]interface{}, history []HistoryEvent, retryPolicy RetryPolicy) *WorkflowRunner {
	return &WorkflowRunner{
		workflowFunc:       workflowFunc,
		inputPayload:       inputPayload,
		historyView:        NewHistoryView(history),
		signalConsumed:     make(map[string]int),
		defaultRetryPolicy: retryPolicy,
	}
}

func (r *WorkflowRunner) Run() (result WorkflowRunResult) {
	defer func() {
		if rec := recover(); rec != nil {
			if yield, ok := rec.(workflowYield); ok {
				result = WorkflowRunResult{Status: yield.status, Command: yield.command}
				return
			}
			result = WorkflowRunResult{Status: "FAILED", Error: fmt.Sprintf("%v", rec)}
		}
	}()

	ctx := &WorkflowContext{runner: r}
	output, err := r.workflowFunc(ctx, r.inputPayload)
	if err != nil {
		return WorkflowRunResult{Status: "FAILED", Error: err.Error()}
	}
	return WorkflowRunResult{Status: "COMPLETED", Result: output}
}

func (r *WorkflowRunner) handleActivity(name string, input map[string]interface{}, opts *ActivityOptions) (interface{}, error) {
	if opts == nil {
		opts = &ActivityOptions{
			ScheduleToCloseTimeoutSeconds: 300,
			StartToCloseTimeoutSeconds:    60,
			HeartbeatTimeoutSeconds:       30,
		}
	}
	if opts.ScheduleToCloseTimeoutSeconds == 0 {
		opts.ScheduleToCloseTimeoutSeconds = 300
	}
	if opts.StartToCloseTimeoutSeconds == 0 {
		opts.StartToCloseTimeoutSeconds = 60
	}
	if opts.HeartbeatTimeoutSeconds == 0 {
		opts.HeartbeatTimeoutSeconds = 30
	}
	retryPolicy := r.defaultRetryPolicy
	if opts.RetryPolicy != nil {
		retryPolicy = *opts.RetryPolicy
	}
	if r.activityIndex < len(r.historyView.ActivityScheduled) {
		// Replay path: the activity is already scheduled in history.
		scheduled := r.historyView.ActivityScheduled[r.activityIndex]
		r.activityIndex++
		expectedName, _ := scheduled.Attributes["activity_name"].(string)
		if expectedName != name {
			return nil, fmt.Errorf("non-determinism: activity name mismatch %s vs %s", expectedName, name)
		}
		activityID, _ := scheduled.Attributes["activity_id"].(string)
		completion, ok := r.historyView.ActivityResults[activityID]
		if !ok {
			// Activity exists but hasn't completed yet. Suspend the workflow task.
			panic(workflowYield{status: "WAITING"})
		}
		switch completion.EventType {
		case EventActivityTaskCompleted:
			return completion.Attributes["result"], nil
		case EventActivityTaskFailed, EventActivityTaskTimedOut:
			return nil, ActivityError{
				Message: fmt.Sprintf("activity failed: %v", completion.Attributes["error"]),
				Details: completion.Attributes,
			}
		default:
			return nil, fmt.Errorf("unexpected completion event %s", completion.EventType)
		}
	}

	// No scheduled activity in history at this point. Emit a new command.
	panic(workflowYield{
		status: "COMMAND",
		command: &Command{
			CommandType: "activity",
			Attributes: map[string]interface{}{
				"activity_name":                     name,
				"input":                             input,
				"retry_policy":                      retryPolicy,
				"schedule_to_close_timeout_seconds": opts.ScheduleToCloseTimeoutSeconds,
				"start_to_close_timeout_seconds":    opts.StartToCloseTimeoutSeconds,
				"heartbeat_timeout_seconds":         opts.HeartbeatTimeoutSeconds,
			},
		},
	})
}

func (r *WorkflowRunner) handleTimer(durationSeconds int) (interface{}, error) {
	if r.timerIndex < len(r.historyView.TimerScheduled) {
		// Replay path: timer already exists in history.
		scheduled := r.historyView.TimerScheduled[r.timerIndex]
		r.timerIndex++
		timerID, _ := scheduled.Attributes["timer_id"].(string)
		if timerID == "" {
			return nil, fmt.Errorf("non-determinism: timer without id")
		}
		if _, ok := r.historyView.TimerResults[timerID]; !ok {
			// Timer scheduled but not fired yet; suspend.
			panic(workflowYield{status: "WAITING"})
		}
		return map[string]interface{}{"timer_id": timerID}, nil
	}
	// No timer yet; emit a new timer command.
	panic(workflowYield{
		status: "COMMAND",
		command: &Command{
			CommandType: "timer",
			Attributes: map[string]interface{}{
				"duration_seconds": durationSeconds,
			},
		},
	})
}

func (r *WorkflowRunner) handleSignal(name string) (map[string]interface{}, error) {
	consumed := r.signalConsumed[name]
	signals := r.historyView.Signals[name]
	if consumed < len(signals) {
		payload := signals[consumed]
		r.signalConsumed[name] = consumed + 1
		return payload, nil
	}
	// Signals are buffered; if none available, suspend deterministically.
	panic(workflowYield{status: "WAITING"})
}

func (r *WorkflowRunner) handleVersion(changeID string, minSupported, maxSupported int) (int, error) {
	if version, ok := r.historyView.VersionMarkers[changeID]; ok {
		if version < minSupported || version > maxSupported {
			return 0, fmt.Errorf("version %d outside supported range for %s", version, changeID)
		}
		return version, nil
	}
	panic(workflowYield{
		status: "COMMAND",
		command: &Command{
			CommandType: "version_marker",
			Attributes: map[string]interface{}{
				"change_id": changeID,
				"version":   maxSupported,
			},
		},
	})
}

File: python/engine/runtime.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from .models import (
    EVENT_ACTIVITY_TASK_COMPLETED,
    EVENT_ACTIVITY_TASK_FAILED,
    EVENT_ACTIVITY_TASK_SCHEDULED,
    EVENT_ACTIVITY_TASK_TIMED_OUT,
    EVENT_SIGNAL_RECEIVED,
    EVENT_TIMER_FIRED,
    EVENT_TIMER_SCHEDULED,
    EVENT_VERSION_MARKER_RECORDED,
    HistoryEvent,
    RetryPolicy,
)


class WorkflowError(Exception):
    pass


class ActivityFailedError(WorkflowError):
    def __init__(self, message: str, details: Optional[Dict[str, Any]] = None) -> None:
        super().__init__(message)
        self.details = details or {}


class NonDeterministicError(WorkflowError):
    pass


@dataclass(frozen=True)
class ActivityCall:
    name: str
    input: Dict[str, Any]
    retry_policy: RetryPolicy
    schedule_to_close_timeout_seconds: int
    start_to_close_timeout_seconds: int
    heartbeat_timeout_seconds: int


@dataclass(frozen=True)
class SleepCall:
    duration_seconds: int


@dataclass(frozen=True)
class WaitSignal:
    signal_name: str


@dataclass(frozen=True)
class VersionMarker:
    change_id: str
    min_supported: int
    max_supported: int


@dataclass
class Command:
    command_type: str
    attributes: Dict[str, Any]


@dataclass
class WorkflowRunResult:
    status: str
    result: Optional[Any] = None
    error: Optional[str] = None
    command: Optional[Command] = None


class WorkflowContext:
    """
    Helper for workflow authoring. All methods return an Operation that must be yielded
    from the workflow generator.
    """

    def __init__(self, default_retry_policy: RetryPolicy) -> None:
        self._default_retry_policy = default_retry_policy

    def execute_activity(
        self,
        name: str,
        input_payload: Dict[str, Any],
        retry_policy: Optional[RetryPolicy] = None,
        schedule_to_close_timeout_seconds: int = 300,
        start_to_close_timeout_seconds: int = 60,
        heartbeat_timeout_seconds: int = 30,
    ) -> ActivityCall:
        return ActivityCall(
            name=name,
            input=input_payload,
            retry_policy=retry_policy or self._default_retry_policy,
            schedule_to_close_timeout_seconds=schedule_to_close_timeout_seconds,
            start_to_close_timeout_seconds=start_to_close_timeout_seconds,
            heartbeat_timeout_seconds=heartbeat_timeout_seconds,
        )

    def sleep(self, duration_seconds: int) -> SleepCall:
        return SleepCall(duration_seconds=duration_seconds)

    def wait_signal(self, signal_name: str) -> WaitSignal:
        return WaitSignal(signal_name=signal_name)

    def get_version(self, change_id: str, min_supported: int, max_supported: int) -> VersionMarker:
        return VersionMarker(
            change_id=change_id,
            min_supported=min_supported,
            max_supported=max_supported,
        )


class HistoryView:
    """
    Pre-processed view of the history to drive deterministic replay.
    """

    def __init__(self, history: List[HistoryEvent]) -> None:
        self.activity_scheduled: List[HistoryEvent] = []
        self.activity_results: Dict[str, HistoryEvent] = {}
        self.timer_scheduled: List[HistoryEvent] = []
        self.timer_results: Dict[str, HistoryEvent] = {}
        self.signals: Dict[str, List[Dict[str, Any]]] = {}
        self.version_markers: Dict[str, int] = {}

        for event in history:
            if event.event_type == EVENT_ACTIVITY_TASK_SCHEDULED:
                self.activity_scheduled.append(event)
            elif event.event_type in (
                EVENT_ACTIVITY_TASK_COMPLETED,
                EVENT_ACTIVITY_TASK_FAILED,
                EVENT_ACTIVITY_TASK_TIMED_OUT,
            ):
                activity_id = event.attributes.get("activity_id")
                if activity_id:
                    self.activity_results[activity_id] = event
            elif event.event_type == EVENT_TIMER_SCHEDULED:
                self.timer_scheduled.append(event)
            elif event.event_type == EVENT_TIMER_FIRED:
                timer_id = event.attributes.get("timer_id")
                if timer_id:
                    self.timer_results[timer_id] = event
            elif event.event_type == EVENT_SIGNAL_RECEIVED:
                name = event.attributes.get("signal_name")
                payload = event.attributes.get("payload", {})
                if name:
                    self.signals.setdefault(name, []).append(payload)
            elif event.event_type == EVENT_VERSION_MARKER_RECORDED:
                change_id = event.attributes.get("change_id")
                version = event.attributes.get("version")
                if change_id is not None and version is not None:
                    self.version_markers[str(change_id)] = int(version)


class WorkflowRunner:
    """
    Deterministic workflow runner that replays history and returns either a new command,
    a waiting state, or final completion.
    """

    def __init__(
        self,
        workflow_func,
        input_payload: Dict[str, Any],
        history: List[HistoryEvent],
        default_retry_policy: RetryPolicy,
    ) -> None:
        self._workflow_func = workflow_func
        self._input = input_payload
        self._history_view = HistoryView(history)
        self._activity_index = 0
        self._timer_index = 0
        self._signal_consumed: Dict[str, int] = {}
        self._default_retry_policy = default_retry_policy

    def run(self) -> WorkflowRunResult:
        context = WorkflowContext(self._default_retry_policy)
        generator = self._workflow_func(context, self._input)
        send_value = None

        try:
            while True:
                try:
                    if send_value is None:
                        operation = next(generator)
                    else:
                        operation = generator.send(send_value)
                    send_value = None
                except StopIteration as exc:
                    return WorkflowRunResult(status="COMPLETED", result=exc.value)

                if isinstance(operation, ActivityCall):
                    result = self._handle_activity(operation)
                    if result.status == "COMMAND":
                        return result
                    if result.status == "WAITING":
                        return result
                    if result.status == "ERROR":
                        try:
                            send_value = generator.throw(
                                ActivityFailedError(result.error or "activity failed", result.result)
                            )
                        except StopIteration as exc:
                            return WorkflowRunResult(status="COMPLETED", result=exc.value)
                    else:
                        send_value = result.result
                elif isinstance(operation, SleepCall):
                    result = self._handle_timer(operation)
                    if result.status == "COMMAND":
                        return result
                    if result.status == "WAITING":
                        return result
                    send_value = result.result
                elif isinstance(operation, WaitSignal):
                    result = self._handle_signal(operation)
                    if result.status == "WAITING":
                        return result
                    send_value = result.result
                elif isinstance(operation, VersionMarker):
                    result = self._handle_version_marker(operation)
                    if result.status == "COMMAND":
                        return result
                    send_value = result.result
                else:
                    raise NonDeterministicError(
                        f"Unsupported operation type {type(operation).__name__}"
                    )
        except Exception as exc:
            return WorkflowRunResult(status="FAILED", error=str(exc))

    def _handle_activity(self, operation: ActivityCall) -> WorkflowRunResult:
        # Replay path: if the activity has already been scheduled in history, we
        # must NOT schedule it again. Instead we look for its completion event.
        if self._activity_index < len(self._history_view.activity_scheduled):
            scheduled_event = self._history_view.activity_scheduled[self._activity_index]
            self._activity_index += 1
            if scheduled_event.attributes.get("activity_name") != operation.name:
                raise NonDeterministicError(
                    f"Activity name mismatch. Expected {scheduled_event.attributes.get('activity_name')} "
                    f"but got {operation.name}"
                )
            activity_id = scheduled_event.attributes.get("activity_id")
            if not activity_id:
                raise NonDeterministicError("Scheduled activity missing activity_id")
            completion = self._history_view.activity_results.get(activity_id)
            if completion is None:
                # The activity is scheduled but has not completed yet. We must
                # suspend the workflow task and wait for the completion event.
                return WorkflowRunResult(status="WAITING")
            if completion.event_type == EVENT_ACTIVITY_TASK_COMPLETED:
                return WorkflowRunResult(status="RESULT", result=completion.attributes.get("result"))
            error_message = completion.attributes.get("error", "activity failed")
            return WorkflowRunResult(status="ERROR", error=error_message, result=completion.attributes)

        # No scheduled activity exists in history at this point, so emit a new
        # command for the engine to persist and schedule.
        return WorkflowRunResult(
            status="COMMAND",
            command=Command(
                command_type="activity",
                attributes={
                    "activity_name": operation.name,
                    "input": operation.input,
                    "retry_policy": operation.retry_policy.to_dict(),
                    "schedule_to_close_timeout_seconds": operation.schedule_to_close_timeout_seconds,
                    "start_to_close_timeout_seconds": operation.start_to_close_timeout_seconds,
                    "heartbeat_timeout_seconds": operation.heartbeat_timeout_seconds,
                },
            ),
        )

    def _handle_timer(self, operation: SleepCall) -> WorkflowRunResult:
        # Replay path: if the timer is already scheduled, check for its firing.
        if self._timer_index < len(self._history_view.timer_scheduled):
            scheduled_event = self._history_view.timer_scheduled[self._timer_index]
            self._timer_index += 1
            timer_id = scheduled_event.attributes.get("timer_id")
            if not timer_id:
                raise NonDeterministicError("Timer scheduled without timer_id")
            fired = self._history_view.timer_results.get(timer_id)
            if fired is None:
                # Timer is scheduled but not fired. Wait until scheduler fires it.
                return WorkflowRunResult(status="WAITING")
            return WorkflowRunResult(status="RESULT", result={"timer_id": timer_id})

        # No timer exists in history yet; emit a new timer command.
        return WorkflowRunResult(
            status="COMMAND",
            command=Command(
                command_type="timer",
                attributes={"duration_seconds": operation.duration_seconds},
            ),
        )

    def _handle_signal(self, operation: WaitSignal) -> WorkflowRunResult:
        name = operation.signal_name
        consumed = self._signal_consumed.get(name, 0)
        available = self._history_view.signals.get(name, [])
        if consumed < len(available):
            payload = available[consumed]
            self._signal_consumed[name] = consumed + 1
            return WorkflowRunResult(status="RESULT", result=payload)
        # Signals are buffered in history; if none are available yet, block.
        return WorkflowRunResult(status="WAITING")

    def _handle_version_marker(self, operation: VersionMarker) -> WorkflowRunResult:
        if operation.change_id in self._history_view.version_markers:
            version = self._history_view.version_markers[operation.change_id]
            if not (operation.min_supported <= version <= operation.max_supported):
                raise NonDeterministicError(
                    f"Version {version} for change_id {operation.change_id} outside supported range"
                )
            return WorkflowRunResult(status="RESULT", result=version)
        return WorkflowRunResult(
            status="COMMAND",
            command=Command(
                command_type="version_marker",
                attributes={
                    "change_id": operation.change_id,
                    "version": operation.max_supported,
                },
            ),
        )

